tool-openssl/pkcs8.cc

```diff
@@ -0,0 +1,262 @@
+  { "-inform", kOptionalArgument, "Input format (DER or PEM)" },
+  { "-outform", kOptionalArgument, "Output format (DER or PEM)" },
+  { "-topk8", kBooleanArgument, "Convert traditional format to PKCS#8" },
+  { "-nocrypt", kBooleanArgument, "Use unencrypted private key" },
+  { "-v2", kOptionalArgument, "Use PKCS#5 v2.0 and specified cipher" },
+  { "-v2prf", kOptionalArgument, "Use specified PRF algorithm with PKCS#5 v2.0" },
+  { "-passin", kOptionalArgument, "Input file passphrase source" },
+  { "-passout", kOptionalArgument, "Output file passphrase source" },
+  { "", kOptionalArgument, "" }
+};
+
+bool pkcs8Tool(const args_list_t &args) {
+  args_map_t parsed_args;
+  args_list_t extra_args;
+  if (!ParseKeyValueArguments(parsed_args, extra_args, args, kArguments) ||
+      extra_args.size() > 0) {
+    PrintUsage(kArguments);
+    return false;
+  }
+
+  bool help = false;
+  GetBoolArgument(&help, "-help", parsed_args);
+  if (help) {
+    PrintUsage(kArguments);
+    return false;
+  }
+
+  // Parse options
+  std::string in_path, out_path, inform, outform;
+  std::string v2_cipher, v2prf;
+  std::string passin_arg, passout_arg;
+  bool topk8 = false, nocrypt = false;
+
+  GetString(&in_path, "-in", "", parsed_args);
+  GetString(&out_path, "-out", "", parsed_args);
+  GetString(&inform, "-inform", "PEM", parsed_args);
+  GetString(&outform, "-outform", "PEM", parsed_args);
+  GetBoolArgument(&topk8, "-topk8", parsed_args);
+  GetBoolArgument(&nocrypt, "-nocrypt", parsed_args);
+  GetString(&v2_cipher, "-v2", "", parsed_args);
+  GetString(&v2prf, "-v2prf", "", parsed_args);
+  GetString(&passin_arg, "-passin", "", parsed_args);
+  GetString(&passout_arg, "-passout", "", parsed_args);
+
+  // Check required arguments
+  if (in_path.empty()) {
+    fprintf(stderr, "Error: missing required argument '-in'\n");
+    return false;
+  }
+
+  // Convert format strings to uppercase for consistent comparison
+  for (auto &c : inform) c = toupper(c);
+  for (auto &c : outform) c = toupper(c);
+
+  // Validate formats
+  if (inform != "PEM" && inform != "DER") {
+    fprintf(stderr, "Error: Invalid input format. Must be PEM or DER.\n");
+    return false;
+  }
+  if (outform != "PEM" && outform != "DER") {
+    fprintf(stderr, "Error: Invalid output format. Must be PEM or DER.\n");
+    return false;
+  }
+
+  // Open input file
+  ScopedFILE in_file(fopen(in_path.c_str(), "rb"));
+  if (!in_file) {
+    fprintf(stderr, "Error: unable to open input file '%s'\n", in_path.c_str());
+    return false;
+  }
+
+  // Prepare output file
+  ScopedFILE out_file;
+  if (!out_path.empty()) {
+    out_file.reset(fopen(out_path.c_str(), "wb"));
+    if (!out_file) {
+      fprintf(stderr, "Error: unable to open output file '%s'\n", out_path.c_str());
+      return false;
+    }
+  }
+  FILE *out = out_file.get() ? out_file.get() : stdout;
+
+  // Get passwords if needed
+  char *passin = nullptr, *passout = nullptr;
```

Comment:
We should be able to use a smart pointer for both variables here. That way we can avoid the additional `OPENSSL_free` calls everywhere and get C++ to do the work for us.